src/compiler/scala/tools/nsc/ast/parser/Scanners.scala

```diff
@@ -1087,34 +1091,29 @@ trait Scanners extends ScannersCommon {
       }
     }
 
-    def doubleVal: Double = doubleVal(negated = false)
+    @`inline` def doubleVal: Double = doubleVal(negated = false)
 
-    def checkNoLetter(): Unit = {
-      if (isIdentifierPart(ch) && ch >= ' ')
-        syntaxError("Invalid literal number")
-    }
+    @`inline` def checkNoLetter(): Unit = if (isIdentifierPart(ch) && ch >= ' ') syntaxError("invalid literal number")
+
+    @`inline` private def isNumberSeparator(c: Char): Boolean = c == '_'
 
-    @inline private def isNumberSeparator(c: Char): Boolean = c == '_'
+    @`inline` private def removeNumberSeparators(s: String): String = if (s.indexOf('_') > 0) s.replace("_", "") else s
 
-    @inline private def removeNumberSeparators(s: String): String =
-      if (s.indexOf('_') > 0) s.replace("_", "") else s
+    @`inline` private def numberOffset = offset + (if (base == 10) 0 else 2)
 
-    // disallow trailing numeric separator char, but let lexing limp along
+    // disallow trailing numeric separator char
     def checkNoTrailingSeparator(): Unit =
-      if (!cbuf.isEmpty && isNumberSeparator(cbuf.last)) {
-        syntaxError(offset + cbuf.length - 1, "trailing separator is not allowed")
-        cbuf.setLength(cbuf.length - 1)
-      }
+      if (!cbuf.isEmpty && isNumberSeparator(cbuf.last))
+        syntaxError(numberOffset + cbuf.length - 1, "illegal separator")
 
     /** Read a number into strVal.
      *
-     *  The `base` can be 8, 10 or 16, where base 8 flags a leading zero.
-     *  For ints, base 8 is legal only for the case of exactly one zero.
+     *  The `base` can be 2, 10 or 16.
```

Comment:
Trying to fix the `0x == 0` bug in dotty and was confused looking at current scala 2 because this isn't merged yet.